docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+    - `9 + (\a b -> a + b) 1 2`
+- inline it only after at least one argument has been applied to it (i.e. the directive is `arity=1`)
+    - `9 + (\  b -> 1 + b)   2`
+- inline it only after at least two arguments have been applied to it (i.e. the directive is `arity=2`)
+    - `9 + (        1 + 2)`
+
+**To summarize, inline directives provide developers with a stronger guarantee about whether and when a function is inlined than just relying upon the analysis done by `purs-backend-es`.**
+
+#### Inlining Duplicates Code
+
+The below example highlights the risk of inlining everything without thought: pointless code duplication.
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = hasExpensiveLetBinding 1 2
+usage2 = hasExpensiveLetBinding 3 4
+```
+
+If we inline `hasExpensiveLetBinding` immediately, the `expensiveComputation` will be computed three times rather than once. In other words, it would be the same as writing the following in source code:
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 1 2
+usage2 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 3 4
+```
+
+**To summarize, inlining for the sake of inlining is a great way to unnecessarily increase your program's bundle size**.
+
+#### Rewrite Rules
+
+A **rewrite rule** will replace one expression, usually involving literal values, with an equivalent expression. For example, an expression like `1 + 2` will always produce the value `3`. So, we can replace that expression with `3`. The rewrite rule here could be defined as, "Whenever the `+` function is called on two literal `Int` values, replace that expression with the sum of the two `Int` values."
+
+```purs
+-- Before
+foo = 1 + 2
+  --  ^^^^^ Rewrite rule: "I can optimize that! `1 + 2` is `3`"
+
+-- After
+foo = 3
+```
+
+However, some expressions will prevent such rewrite rules from triggering. For example:
+
+```purs
+foo =
+  let
+    a = { bar: 42 }
+  in
+    1 + a.bar
+--  ^^^^^^^^^ Rewrite rule: "I can't optimize that! `a.bar` is not a literal `Int` value."
+```
+
+Because the rewrite rule cannot see the literal `Int` value represented by `a.bar`, its optimization does not trigger. To solve this problem, we need to make it visible. The solution is to duplicate code via inlining.
+
+**To summarize, inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes.**
+
+#### Code Optimzation Example via Rewrite Rules and Inlining
+
+Let's see how this works in practice using this example:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = ignoreArgs 8 9
+```
```

Comment:
This whole series of steps is not quite accurate because the optimizer will optimizer ignoreArgs on it's own to
```
ignoreArgs arg1 arg2 =
  42 + arg1 + arg2
```
Before it even does anything with `foo`.